### PR TITLE
feat: refine operator user credit ledger filters

### PR DESCRIPTION
**Pull request by QinyingShi**
Summary
- add dynamic operator user credit ledger filters for type, grant source, usage mode, course, and time range
- merge consume course search into a single ID/name field and refine the tab card layout to match the course detail page
- align user and course credit usage copy for read/follow-up scenarios and cover the new filter behavior with backend/frontend tests

Testing
- `cd src/cook-web && npm test -- --runInBand --runTestsByPath 'src/app/admin/operations/users/[user_bid]/page.test.tsx'`
- `cd src/cook-web && npm run type-check`
- `cd src/api && pytest tests/service/shifu/test_admin_users.py -q`
- `pre-commit run -a`

## Summary by CodeRabbit

* **New Features**
  * Credit Ledger tab: interactive filters (credit type, grant source, course/query, usage mode, date range), paginated/searchable ledger list, improved display metadata and timestamps.

* **Chores**
  * Backend and API docs extended to support the new ledger filters; client-side filter types and utilities added; UI wiring for filter state and handlers.

* **Tests**
  * Expanded tests covering consume/grant/other filters, course/usage-mode searches, time windowing, and edge cases.

[Review Change Stack](https://app.coderabbit.ai/change-stack/ai-shifu/ai-shifu/pull/1762)

### CODE REVIEW

**Comment by QinyingShi** [comment]
Addressed the review follow-ups in the latest updates:

- restored the DB-paginated fast path for operator user credit ledger requests when no Python-only sub-filter is active
- kept consume ledger rows visible when the related usage record is missing unless a course / usage-mode sub-filter actually requires that lookup
- extracted the user credit ledger tab UI and filter sanitization out of the route entry into local modules
- deduplicated usage lookup ids before building the `IN` clause

Validation rerun:
- `cd src/cook-web && npm test -- --runInBand --runTestsByPath 'src/app/admin/operations/users/[user_bid]/page.test.tsx'`
- `cd src/cook-web && npm run type-check`
- `cd src/api && pytest tests/service/shifu/test_admin_users.py -q`
- `pre-commit run -a`

**Comment by QinyingShi** [comment]
Addressed the review follow-ups in the latest updates:\n\n- restored the DB-paginated fast path for operator user credit ledger requests when no Python-only sub-filter is active\n- kept consume ledger rows visible when the related usage record is missing unless a course / usage-mode sub-filter actually requires that lookup\n- extracted the user credit ledger tab UI and filter sanitization out of the route entry into local modules\n- deduplicated usage lookup ids before building the  clause\n\nValidation rerun:\n- 
> cook-web@2.0.0 test
> jest --runInBand --runTestsByPath src/app/admin/operations/users/[user_bid]/page.test.tsx\n- 
> cook-web@2.0.0 type-check
> tsc --noEmit\n- ...........................................                              [100%]
=============================== warnings summary ===============================
../../../../../Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/pydantic/fields.py:814: 306 warnings
tests/service/shifu/test_admin_users.py: 263 warnings
  /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/pydantic/fields.py:814: PydanticDeprecatedSince20: Using extra keyword arguments on `Field` is deprecated and will be removed. Use `json_schema_extra` instead. (Extra keys: 'required'). Deprecated in Pydantic V2.0 to be removed in V3.0. See Pydantic V2 Migration Guide at https://errors.pydantic.dev/2.8/migration/
    warn(

tests/service/shifu/test_admin_users.py::test_list_operator_users_returns_paginated_summaries_with_resolved_metadata
  /Library/Frameworks/Python.framework/Versions/3.11/lib/python3.11/site-packages/pydantic/_internal/_config.py:291: PydanticDeprecatedSince20: Support for class-based `config` is deprecated, use ConfigDict instead. Deprecated in Pydantic V2.0 to be removed in V3.0. See Pydantic V2 Migration Guide at https://errors.pydantic.dev/2.8/migration/
    warnings.warn(DEPRECATION_MESSAGE, DeprecationWarning)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
43 passed, 570 warnings in 6.89s\n- check yaml...............................................................Passed
check json...............................................................Passed
check for merge conflicts................................................Passed
check for broken symlinks................................................Passed
fix end of files.........................................................Passed
trim trailing whitespace.................................................Passed
pretty format json.......................................................Passed
fix requirements.txt.................................(no files to check)Skipped
ruff check...............................................................Passed
ruff format..............................................................Passed
Prettier cook-web files..................................................Passed
Validate repository harness..............................................Passed
Validate architecture boundaries.........................................Passed
Validate shared translation parity.......................................Passed
Validate translation key usage...........................................Passed
Validate locales metadata up to date.....................................Passed
ESLint cook-web files (errors only)......................................Passed
Forbid hardcoded CJK characters in backend Python........................Passed

**Comment by QinyingShi** [comment]
Addressed the latest follow-ups in `1d698d90`:

- moved the remaining `grant_source` and consume sub-filters onto SQL-bounded `count + offset + limit` paths, including SQL-side usage-mode filtering and course-title fuzzy matching
- documented the legacy `course_id` / `course_name` compatibility params and their precedence behind `course_query`
- localized the ledger label fallback path and removed hardcoded pagination fallback copy from the extracted ledger tab

Kept the current `creditType=all` time-filter behavior unchanged on purpose: the current product requirement for this page is to show only the type selector by default, then reveal the type-specific filters after a type is chosen.

Validation rerun:
- `cd src/api && pytest tests/service/shifu/test_admin_users.py -q`
- `cd src/cook-web && npm test -- --runInBand --runTestsByPath 'src/app/admin/operations/users/[user_bid]/page.test.tsx'`
- `cd src/cook-web && npm run type-check`
- `pre-commit run -a`

File: src/api/flaskr/service/shifu/admin.py [hunk 1]
```diff
@@ -5233,6 +5324,105 @@ def grant_operator_user_credits(
...
+        db.session.query(
+            BillingOrder.bill_order_bid.label("bill_order_bid"),
+            db.func.max(BillingOrder.id).label("max_id"),
+        )
+        .filter(BillingOrder.deleted == 0)
+        .group_by(BillingOrder.bill_order_bid)
+        .subquery()
+    )
+
+
```
[the remark this answers is not in the document]
Comment (QinyingShi) [reply src/api/flaskr/service/shifu/admin.py]: Fixed in `a3989ceb`.

I scoped both latest-record subqueries to the current operator target user instead of grouping across the full `bill_usage` / `billing_orders` tables. That keeps the result shape the same for this page, but avoids a full-table aggregate scan for each user credit detail request.

File: src/api/flaskr/service/shifu/admin.py [hunk 2]
```diff
@@ -5331,26 +5523,171 @@ def get_operator_user_credits(
...
+                matching_course_bids = _find_matching_operator_course_bids_by_name(
+                    candidate_shifu_bids,
+                    resolved_course_query,
+                )
+                course_filters = [usage_row.shifu_bid == resolved_course_query]
+                if matching_course_bids:
+                    course_filters.append(
+                        usage_row.shifu_bid.in_(sorted(matching_course_bids))
+                    )
+                query = query.filter(or_(*course_filters))
```
[the remark this answers is not in the document]
Comment (QinyingShi) [reply src/api/flaskr/service/shifu/admin.py]: Fixed in `a3989ceb`.

I removed the extra `query.with_entities(...).distinct().all()` round trip from the consume branch. `course_query` now applies as a direct course filter: exact `shifu_bid` match plus title-based matched course bids loaded from the course tables, without executing the partially built ledger query an extra time.

File: src/api/flaskr/service/shifu/admin.py [hunk 3]
```diff
@@ -5331,26 +5523,171 @@ def get_operator_user_credits(
...
+                        CreditLedgerEntry.source_type == CREDIT_SOURCE_TYPE_SUBSCRIPTION
+                    )
+                )
+                if (
+                    grant_source
+                    == OPERATOR_USER_CREDIT_FILTER_GRANT_SOURCE_TRIAL_SUBSCRIPTION
+                ):
+                    query = query.filter(checkout_type_expr == "trial_bootstrap")
+                else:
+                    query = query.filter(checkout_type_expr != "trial_bootstrap")
```
[the remark this answers is not in the document]
Comment (QinyingShi) [reply src/api/flaskr/service/shifu/admin.py:5645]: This fallback is intentional.

For operator filtering we currently split `trial_bootstrap` out as a dedicated bucket and keep the remaining subscription-source grants under the regular `subscription` bucket, including legacy rows whose `checkout_type` was not fully populated. That preserves historical visibility instead of silently dropping older subscription grants from the operator view.

I agree this behavior is worth making explicit, so I will keep this classification and handle any further clarification as follow-up cleanup instead of changing the bucket semantics in this PR.

File: src/api/flaskr/service/shifu/admin.py [hunk 4]
```diff
@@ -5331,26 +5523,171 @@ def get_operator_user_credits(
...
+            elif usage_mode == COURSE_CREDIT_USAGE_MODE_ASK:
+                query = query.filter(
+                    usage_row.usage_type != BILL_USAGE_TYPE_TTS,
+                    _build_course_credit_usage_ask_filter(generation_name_expr),
+                )
+            elif usage_mode == COURSE_CREDIT_USAGE_MODE_LEARN:
+                query = query.filter(
+                    usage_row.usage_type != BILL_USAGE_TYPE_TTS,
+                    _build_course_credit_usage_learn_filter(generation_name_expr),
+                )
```
[the remark this answers is not in the document]
Comment (QinyingShi) [reply src/api/flaskr/service/shifu/admin.py:5677]: This behavior is intentional for the current filter semantics.

The unfiltered consume view still keeps consume ledger rows even when the backing usage record is missing, and we have coverage for that case. Once the operator applies `course_query` or `usage_mode`, the query needs usage-derived fields to classify the row, so rows without a resolvable usage record are excluded from that narrowed view.

I agree this is important product behavior to call out, but I do not want to change that filtering contract inside this PR.

File: src/api/flaskr/service/shifu/admin.py [hunk 5]
```diff
@@ -754,6 +776,76 @@ def _resolve_operator_credit_note_code(
...
+    if normalized in {"", *OPERATOR_USER_CREDIT_FILTER_TYPES}:
+        return normalized or OPERATOR_USER_CREDIT_TYPE_ALL
+    return ""
+
+
+def _resolve_operator_user_credit_grant_source_filter(value: str) -> str:
+    normalized = str(value or "").strip().lower()
+    if normalized in {"", *OPERATOR_USER_CREDIT_FILTER_GRANT_SOURCES}:
+        return normalized or OPERATOR_USER_CREDIT_FILTER_GRANT_SOURCE_ALL
+    return ""
```
[the remark this answers is not in the document]
Comment (QinyingShi) [reply src/api/flaskr/service/shifu/admin.py:790]: Good call-out.

The empty-value normalization is intentionally different today: `credit_type` uses `all` as an explicit UI/default bucket, while `usage_mode` uses an empty value to mean 'no consume sub-filter applied'. The downstream validation only rejects non-empty invalid input, so the current flow is consistent, but I agree the asymmetry can be surprising when reading the helpers in isolation.

I prefer to keep the current semantics in this PR to avoid broadening the filter contract, and treat alignment / helper cleanup as follow-up work instead.

File: src/api/flaskr/service/shifu/route.py [hunk 6]
```diff
@@ -1367,6 +1367,46 @@ def admin_operation_user_credits(user_bid: str):
...
+            - name: course_id
+              in: query
+              type: string
+              required: false
+              description: Legacy exact course ID filter, used only when course_query is empty
+            - name: course_name
+              in: query
+              type: string
+              required: false
+              description: Legacy fuzzy course name filter, used only when course_query and course_id are empty
```
[the remark this answers is not in the document]
Comment (QinyingShi) [reply src/api/flaskr/service/shifu/route.py]: Fixed in `a3989ceb`.

I removed the legacy `course_id` / `course_name` aliases from the route contract and OpenAPI doc so this filter surface stays centered on a single `course_query` parameter.

File: src/cook-web/src/app/admin/operations/users/[user_bid]/creditFilterUtils.ts [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+  }
+  if (filters.creditType === 'other') {
+    return {
+      ...filters,
+      grantSource: FILTER_ALL_OPTION,
+      courseQuery: '',
+      usageMode: FILTER_ALL_OPTION,
+    };
+  }
+  return createUserCreditFilters();
```
[the remark this answers is not in the document]
Comment (QinyingShi) [reply src/cook-web/src/app/admin/operations/users/[user_bid]/creditFilterUtils.ts:38]: This reset is intentional in the current UX.

The time-range picker is only shown for the typed views (`grant`, `consume`, `other`). When the operator switches back to `all`, we clear the type-specific sub-filters together, including the hidden time range, to avoid keeping an invisible active filter in the request state.

If we later want 'all types within the same time range', that should come with an explicit UI change that keeps the time-range control visible in the `all` view as well.
